charmcraft/utils/package.py

```diff
@@ -103,21 +103,25 @@ def get_pip_command(
     binary_packages = get_pypi_packages(binary_deps)
     requirements_packages = get_requirements_file_package_names(*requirements_files)
     all_packages = charm_packages | binary_packages | requirements_packages
-    source_only_packages = sorted(
-        get_package_names(all_packages) - get_package_names(binary_packages)
-    )
-
     non_requirements_packages = sorted(
         exclude_packages(
             set(source_deps) | set(binary_deps),
             excluded=get_package_names(requirements_packages),
         )
     )
 
-    if source_only_packages:
-        no_binary = [f"--no-binary={','.join(source_only_packages)}"]
-    else:
-        no_binary = []
+    if not binary_packages:
+        return [
+            *prefix,
+            "--no-binary=:all:",
+            *(f"--requirement={path}" for path in requirements_files),
+            *non_requirements_packages,
```

Comment:
`non_requirements_package` may include binary dependencies. 
Will it work in such scenario ?

Reply:
`non_requirements_packages` only contains binary dependencies if `binary_packages` is not empty, as `binary_packages` is a 1:1 map of `binary_deps`

Reply:
So you'll have always a combination of `--no-binary` with explicitly allowed binary_packages ?

Reply:
No, `--no-binary=:all` only gets set if there aren't any explictly allowed binary packages.

If any explicitly-allowed binary packages are set, the code path below that selects that list of binary packages gets used instead.

Reply:
Got it, thanks!